singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,340 @@
+   *   the total rounds the calling thread will wait.
+   */
+  private int callingThreadSleepInSeconds = 30;
+
+  /**
+   *  When sender send audit events to Kafka,  it chooses a random partition and if it fails, it
+   *  will choose another GOOD partition, this retry will happen at most numOfPartitionsToTrySending
+   *  times before dropping the event. Note that, this is different from the retry handled by kafka client
+   *  library when sending an event to a certain partition.
+   */
+  private int numOfPartitionsToTrySending = 3;
+
+  /**
+   *  For each partition, track the number of sending failures happened to this partition.
+   */
+  private Map<Integer, AtomicInteger> partitionErrorCounts = new ConcurrentHashMap<>();
+
+  /**
+   * For each event (identified by LoggingAuditHeaders, key of the map), track the partitions that
+   * has been used to send this events. These partitions are aka bad partitions, because it failed
+   * sending out this event. During retry, new partition not belong the bad partitions  is
+   * randomly chosen. When event is send out successfully or dropped, the corresponding entry in
+   * this map is removed.
+   */
+  private Map<LoggingAuditHeaders, Set<Integer>> eventBadPartitionsMap = new ConcurrentHashMap<>();
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               LinkedBlockingDeque<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage , String host, String name){
+     this.topic = config.getTopic();
+     this.queue = queue;
+     this.stage = stage;
+     this.host = host;
+     this.name = name;
+  }
+
+  public KafkaProducer<byte[], byte[]> getKafkaProducer() {
+    return kafkaProducer;
+  }
+
+  public void setKafkaProducer(KafkaProducer<byte[], byte[]> kafkaProducer) {
+    this.kafkaProducer = kafkaProducer;
+  }
+
+
+  public int getAlternatePartition(Set<Integer> badPartitions, int numOfPartitions){
+    int randomPartition = 0;
+    int trials = 10;
+    while(trials > 0){
+      trials -= 1;
+      randomPartition = ThreadLocalRandom.current().nextInt(numOfPartitions);
+      if (!badPartitions.contains(randomPartition)){
+        break;
+      }
+    }
+    return randomPartition;
+  }
+
+  /**
+   *  Sender dequeue LoggingAuditEvents and send out to Kafka cluster. If send out to one partition
+   *  fails, it will choose another partition. For each event, it will try at most
+   *  umOfPartitionsToTrySending times (default 3 times) before dropping this events. Metrics are
+   *  used to track the queue size and usuage, number of events sent out to Kafka successfully, and
+   *  the number of events dropped.
+   */
+  @Override
+  public void run() {
+    LoggingAuditEvent event = null;
+    ProducerRecord<byte[], byte[]> record;
+    byte[] value = null;
+    List<PartitionInfo> partitionInfoList = this.kafkaProducer.partitionsFor(topic);
```

Comment:
Could we refresh the partition info periodically to prevent us from using staled partition info?

Reply:
implements refresh every 5 mins